### PR TITLE
Map proxy

### PR DESCRIPTION
**Pull request by mustafaiman**
Adds HazelcastClient class and connection initiation logic.
Adds map proxy creation.
Adds Map#put operation.
Adds map put test.
Partition number is hard-coded 271. This will be fixed.
Intoduces basic JsonSerialization to use in map put operation.
Removes old tests and proxy codes.
I fixed generated codecs and used them. They can be replaced by generated codecs when generated codecs are ready.

### CODE REVIEW

File: src/proxy/BaseProxy.ts [hunk 1]
```diff
@@ -1,24 +1,31 @@
...
-    //TODO client should be of type {real client type}
-    constructor(client: { serializationService : SerializationService}, serviceName: string, name: string) {
+    constructor(client: HazelcastClient, serviceName: string, name: string) {
         this.client = client;
         this.name = name;
         this.serviceName = serviceName;
     }
 
+    protected invokeWithPartitionId(clientMessage: ClientMessage, key: any): Q.Promise<ClientMessage> {
+        var partitionId: number = Math.abs(this.client.getSerializationService().toData(key).getPartitionHash()) % 271;
```
Comment (asimarslan) [inline src/proxy/BaseProxy.ts]: PartitionService should handle this. Maybe just an empty one can be better to start with.

Comment (mustafaiman) [inline src/proxy/BaseProxy.ts]: I will abstract this one to a dummy partition service for this pr.

File: src/HazelcastClient.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        this.proxyManager = new ProxyManager(this);
+        this.partitionService = new PartitionService(this);
+        this.connectionManager = new ClientConnectionManager(this.config.networkConfig,
+            this.config.groupConfig, this.invocationService);
+    }
+
+    private init(): Q.Promise<HazelcastClient> {
+        var deferred = Q.defer<HazelcastClient>();
+
+        Q.all([
```
Comment (GhostInAMachine) [inline src/HazelcastClient.ts]: Why are you using "all" here?

Comment (mustafaiman) [inline src/HazelcastClient.ts]: Client should wait for all services to be initialised. You can add other services here in future.

File: src/proxy/BaseProxy.ts [hunk 3]
```diff
@@ -1,24 +1,49 @@
...
+        var deferred: Q.Deferred<T> = Q.defer<T>();
+        var toObject = this.toObject.bind(this);
+        promise.then(function(clientMessage: ClientMessage) {
+            var parameters: any = codec.decodeResponse(clientMessage, toObject);
+            deferred.resolve(parameters.response);
+        });
+        return deferred.promise;
+    }
+
+    protected encodeInvokeOnKey<T>(codec: any, key: any, ...codecArguments: any[]): Q.Promise<T> {
```
Comment (GhostInAMachine) [inline src/proxy/BaseProxy.ts]: What is the point of having both invokeOnKey and invokeOnPartition? Key is mapped to partition key anyway.

Also, I think request should be encoded outside of the scope of this method

File: src/proxy/BaseProxy.ts [hunk 4]
```diff
@@ -1,24 +1,49 @@
...
+        var partitionId: number = this.client.getPartitionService().getPartitionId(key);
+        return this.encodeInvokeOnPartition<T>(codec, partitionId, ...codecArguments);
+    }
+
+    protected encodeInvokeOnPartition<T>(codec: any, partitionId: number, ...codecArguments: any[]): Q.Promise<T> {
```
Comment (GhostInAMachine) [inline src/proxy/BaseProxy.ts]: Why can't you encode the request at the proxy level and pass the raw bytes here? Passing codec class with arguments here confuses me really.

File: src/proxy/BaseProxy.ts (hunk 3 above)
Comment (mustafaiman) [inline src/proxy/BaseProxy.ts]: Do you suggest to calculate partition id of a key in proxy code? In that case, we have to put the same calculation code to each method of map(get, put, remove, etc).

Comment (GhostInAMachine) [inline src/proxy/BaseProxy.ts]: No, encode the request at proxy level (e.g. MapProxy), but calculate the partition ID here

Comment (mustafaiman) [inline src/proxy/BaseProxy.ts]: Where encoding requests happen is another issue. I do not think this is related with comment you wrote here.

Comment (GhostInAMachine) [inline src/proxy/BaseProxy.ts]: Sorry, I indeed replied to wrong comment. Basically, you can inline the invokeOnPartition method

Comment (mustafaiman) [inline src/proxy/BaseProxy.ts]: There are operations which are not assigned with a key but are assigned with a partition. You can see these methods exist in other clients, too.

File: src/proxy/BaseProxy.ts (hunk 4 above)
Comment (mustafaiman) [inline src/proxy/BaseProxy.ts]: We discussed this in Slack channel. I thought we agreed to do it this way.

Comment (GhostInAMachine) [inline src/proxy/BaseProxy.ts]: With whom did you discuss?

File: src/codec/MapPutCodec.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import {Data} from '../serialization/Data';
+
+var REQUEST_TYPE = 0x0101;
+var RESPONSE_TYPE = 105;
+
+export class MapPutCodec{
+    static calculateSize(name: string, key: Data, value: Data, threadId: number, ttl: number){
+        // Calculates the request payload size
+        var dataSize = 0;
+        dataSize += SizeUtil.getStringSize(name);
```
Comment (mustafaiman) [inline src/codec/MapPutCodec.ts]: I know this is not related to codec work but can we move `SizeUtil.getStringSize` to `BitsUtil` and change its name to calculateSizeString with this PR.

File: src/HazelcastClient.ts (hunk 2 above)
Comment (GhostInAMachine) [inline src/HazelcastClient.ts]: OK

File: src/proxy/BaseProxy.ts (hunk 3 above)
Comment (GhostInAMachine) [inline src/proxy/BaseProxy.ts]: OK

File: src/proxy/BaseProxy.ts (hunk 4 above)
Comment (GhostInAMachine) [inline src/proxy/BaseProxy.ts]: Issue resolved in Slack

File: src/codec/MapPutCodec.ts (hunk 5 above)
Comment (GhostInAMachine) [inline src/codec/MapPutCodec.ts]: Yeah, I seen no harm in that. SizeUtil was a temporary class of mine while I was waiting for auto-generated codecs to arrive

Comment (mustafaiman) [inline src/codec/MapPutCodec.ts]: Actually, I was trying to comment on Codec PR about this :) Thanks for approval anyway:)

**Comment by GhostInAMachine** [comment]
Looks good

**Comment by asimarslan** [comment]
I couldn't see the Map proxy optional parameters like put ttl etc.. Why did they removed?

**Comment by mustafaiman** [comment]
@asimarslan They are not removed. Since they are optional parameters in `IMap` Typescript does not require to implement all possible signatures of those methods. I implemented only basic put operation for this PR. I will add other functionality in another PR.

**Comment by asimarslan** [comment]
TypeScript have default values for optional parameters. As our codecs requires all parameters, it can be easier to put these into function signature instead of into codec parameter list.
For example put :
```javascript
function put(key: any, value:any, ttl = -1) {
      ..... MapPutCodec, keyData, valueData, ttl, ...
}
```

**Comment by mustafaiman** [comment]
good point. done.

**Comment by asimarslan** [comment]
:+1:

**Pull request merged**